contrib/submit-protocol/src/App/MarketplaceForm.tsx

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import React from "react";
+import {
+  Callout, DefaultButton, Dropdown, DropdownMenuItemType,
+  Fabric, IDropdownOption, Stack, TextField,
+  mergeStyleSets,
+} from "office-ui-fabric-react";
+
+type MarketplaceUriType = (null | "GitHub");
+
+interface IProtocolItem {
+  name: string;
+  uri: string;
+}
+
+interface IMarketplaceProps {
+  defaultURI: string;
+  defaultURIType: MarketplaceUriType;
+  onSelectProtocol: ((text: string) => void);
+  disabled?: boolean;
+}
+
+interface IMarketplaceState {
+  uri: string;
+  uriType: MarketplaceUriType;
+  protocolOptions: IDropdownOption[];
+  uriConfigCallout: boolean;
+}
+
+const styles = mergeStyleSets({
+  dropdown: {
+    width: 240,
+  },
+  textfiled: {
+    width: 480,
+  },
+});
+
+const defaultProtocolOptions: IDropdownOption[] = [
+  { key: "None", text: "None" },
+  { key: "jobsDivider", text: "-", itemType: DropdownMenuItemType.Divider },
+  { key: "jobsHeader", text: "Protocol Jobs", itemType: DropdownMenuItemType.Header },
+];
+
+export default class MarketplaceForm extends React.Component<IMarketplaceProps, IMarketplaceState> {
+  public static defaultProps: Partial<IMarketplaceProps> = {
+    defaultURI: "https://api.github.com/repos/Microsoft/pai/contents/marketplace-v2",
+    defaultURIType: "GitHub" as MarketplaceUriType,
+    disabled: false,
+  };
+
+  public state = {
+    uri: this.props.defaultURI,
+    uriType: this.props.defaultURIType,
+    protocolOptions: [... defaultProtocolOptions],
```

Comment:
No need to clone this array if it is guaranteed to be immutable.

Reply:
updated